internal/docstore/dynamodocstore/dynamo.go

```diff
@@ -445,74 +461,104 @@ func (c *collection) put(ctx context.Context, k driver.ActionKind, doc driver.Do
 	}
 	if c.sortKey != "" && mf == c.sortKey {
 		// It doesn't make sense to generate a random sort key.
-		return fmt.Errorf("missing sort key %q", c.sortKey)
+		return nil, fmt.Errorf("missing sort key %q", c.sortKey)
 	}
 	rev := driver.UniqueString()
 	if av.M[c.opts.RevisionField], err = encodeValue(rev); err != nil {
-		return err
+		return nil, err
 	}
-	in := &dyn.PutItemInput{
+	dput := &dyn.Put{
 		TableName: &c.table,
 		Item:      av.M,
 	}
-	if condition != nil {
-		ce, err := expression.NewBuilder().WithCondition(*condition).Build()
+	cb, err := c.precondition(a)
+	if err != nil {
+		return nil, err
+	}
+	if cb != nil {
+		ce, err := expression.NewBuilder().WithCondition(*cb).Build()
 		if err != nil {
-			return err
+			return nil, err
 		}
-		in.ExpressionAttributeNames = ce.Names()
-		in.ExpressionAttributeValues = ce.Values()
-		in.ConditionExpression = ce.Condition()
-	}
-	_, err = c.db.PutItemWithContext(ctx, in)
-	if err != nil {
-		if ae, ok := err.(awserr.Error); ok && ae.Code() == dyn.ErrCodeConditionalCheckFailedException {
-			if k == driver.Create {
-				return gcerr.Newf(gcerr.AlreadyExists, err, "document already exists")
+		dput.ExpressionAttributeNames = ce.Names()
+		dput.ExpressionAttributeValues = ce.Values()
+		dput.ConditionExpression = ce.Condition()
+	}
+	return &writeOp{
+		action:          a,
+		writeItem:       &dyn.TransactWriteItem{Put: dput},
+		newPartitionKey: newPartitionKey,
+		newRevision:     rev,
+		run: func(ctx context.Context) error {
+			in := &dyn.PutItemInput{
+				TableName:                 dput.TableName,
+				Item:                      dput.Item,
+				ConditionExpression:       dput.ConditionExpression,
+				ExpressionAttributeNames:  dput.ExpressionAttributeNames,
+				ExpressionAttributeValues: dput.ExpressionAttributeValues,
 			}
-			if rev, _ := doc.GetField(c.opts.RevisionField); rev == nil && k == driver.Replace {
-				return gcerr.Newf(gcerr.NotFound, nil, "document not found")
+			_, err := c.db.PutItemWithContext(ctx, in)
+			if err != nil {
```

Comment:
This `if` could be removed and L512 can just `return err`.

Reply:
OK, done in next PR.